### PR TITLE
Feature operators

### PR DESCRIPTION
**Pull request by flbulgarelli**
Releated to #233 

This PR introduces operators to mulang. They allow to: 

* have more fine-grained inspections
* normalize many common primitives - like addition, hash method or function composition - across languages ASTs
* have less language-specific code in inspections modules
* generalize operator-override in object-oriented languages

Also, this PR comes with a new language-dependent module called `Autocorrector` , which: 

* converts ill-formed language-specific inspections into the true generic expectations. E.g. it converts `* Uses:+` into `UsesPlus`
* converts ill-formed inspections that are based on keywords into true generic expectations. E.g it converts `Uses:for` into `UsesForeach` in Ruby
* deduces the language caseStyle for the language even it was not given

Finally, new inspections were introduced: 

* `usesArithmetic`
* `usesBooleanLogic`
* `usesPlus`, `usesMinus`, `usesAnd`, `usesOtherwise`, and so on

Although `Operator` type deprecates the `Equal`, `NotEqual`, `EqualMethod` and `HashMethod`, this PR still keeps those constructors and processes them.

### CODE REVIEW

**Comment by flbulgarelli** [comment]
@julian-berbel please review and merge if everything is ok

File: src/Language/Mulang/Inspector/Generic/Smell.hs [hunk 1]
```diff
@@ -1,7 +1,10 @@
+{-# LANGUAGE ViewPatterns #-}
+
 module Language.Mulang.Inspector.Generic.Smell (
   hasRedundantBooleanComparison,
   hasRedundantIf,
   hasRedundantGuards,
+  shouldUseOtherwise,
```
Comment (julian-berbel) [inline src/Language/Mulang/Inspector/Generic/Smell.hs]: This doesn't follow the rest of the smells' naming scheme. Feel free to disregard

File: src/Language/Mulang/Operators/Haskell.hs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import Language.Mulang.Ast.Operator (Operator (..))
+
+haskellTokensTable :: TokensTable
+haskellTokensTable =
+  buildTokensTable [
+    (NotEqual, ["/="]),
+    (Negation, ["not"]),
+    (Otherwise, ["otherwise"]),
+    (BackwardComposition, ["."]),
+    (Divide, ["/", "div"])
```
Comment (julian-berbel) [inline src/Language/Mulang/Operators/Haskell.hs]: won't this cause ambiguous parsing?

File: src/Language/Mulang/Inspector/Generic/Smell.hs (hunk 1 above)
Comment (flbulgarelli) [reply src/Language/Mulang/Inspector/Generic/Smell.hs]: You are right!

File: src/Language/Mulang/Operators/Haskell.hs (hunk 2 above)
Comment (flbulgarelli) [reply src/Language/Mulang/Operators/Haskell.hs]: Yes, you are right. I will remove the `div` from the list

**Pull request merged**